### PR TITLE
InfluxDB Support

### PR DESCRIPTION
**Pull request by pimeys**
Hi,

This is an early pull request to open an issue and have some conversation about the idea.

I'm a T1 diabetic and getting my own pump and CGM system very soon. Of course I want to hack that! I'm already having an InfluxDB[1] instance with Grafana[2] running on my own server, so naturally I want to host data about my body and medication in there.

InfluxDB is a time series database which suits for this kind of data way better than MongoDB. With Grafana, one can build lots of nice graphs and dashboards from the raw data, which is something I can't wait to start doing.

I don't know do you want this to the master or is it useful to anybody else than me. I tested this with test data (having no system yet) and it works, also with the official InfluxDB library when they release one bugfix I opened[3].

I'm happy to contribute more when I get the system.

[1] https://www.influxdata.com/time-series-platform/influxdb/
[2] http://grafana.org/
[3] https://github.com/influxdata/influxdb-java/pull/273

### CODE REVIEW

**Comment by jamorham** [comment]
Hi there, thanks for submitting this. 

Broadly I am in favor of adding the ability to upload the data to influxdb as far as the project road map goes. The caveat to this is that it must be done with the minimum impact/overhead on users of the app not utilizing this feature.

From an engineering perspective I'm only going to tell you what is wrong with this and suggest how we can fix and resolve those issues so here goes:

1) This adds a dependency on okhttp3 can it work with okhttp2 instead? Is there a significant difference? I don't want the overhead of including both libraries if it can be avoided.

2) There are problems with `BgSendQueue` and `CalibrationSendQueue` what you've done is mimic the existing structure. Unfortunately the existing structure of those classes which doesn't really offer any sensible way to extend the functionality. By changing the model schema this introduces a downgrade incompatibility which we want to completely avoid.

The mechanism for syncing data like that is superseded by the new `UploaderQueue` class which is extensible for new upload methods using a bitfield instead of booleans. As you may have noticed this is currently only used for treatments (preserving the tested legacy method for uploading other data) which is why it wasn't obvious how to solve your implementation another way.

I think the best and quickest way to resolve this is for you to simple delete the changes to `BgSendQueue` and `CalibrationSendQueue` and `DailyIntentService` and just submit the PR which has the nuts and bolts to do the upload itself and then I can combine this with the correct implementation in `UploaderQueue` because its more complicated than it should be due to the existing legacy of those other classes still being in effect.

Please let me know what you think?

**Comment by pimeys** [comment]
Hi,

Thank you for your answer. This is pretty much an idea and I don't want it to be even near the HEAD until I actually get my own system working and I can see where we can go with InfluxDB. Just a nodge that I want to join this and your fork seems to be the most active.

For the technical issues:

1. I'd say I should just take okhttp2 out and replace it with okhttp3. The deal here is, that we've found a bug[1] from the influx library and a quickfix for now is to give it an interceptor like I did here in my pull request. Now the client wants a builder instead of a client, and this is only available in the version three. It's not complex at all to update the whole library, but I think it's a bit rough in my first pull request, before I actually know the project and who's doing it... :)

2. I'll take a look into these. My java is quite rusty and I'm mainly a Clojure/Rust/Scala developer, so please bear with me if I don't notice all the best practices right away. Getting there though...

I'll get back to you when my tight week schedule is done and I have time for my hobby projects. Let's think about merging this later on, if we think this feature is a good idea in your fork.

[1] https://github.com/influxdata/influxdb-java/issues/268

**Comment by jamorham** [comment]
Do you have any objection to me using what you've done so far and just "making it work" ? I've set up a local influx server to test it with and am interested how much extra it will add to the apk binary.

I didn't previously update to okhttp3 because I wasn't sure if there would be any compatibility issues with it and the various services it is used for within the app and didn't want to break anything which has already widely been tested with okhttp2

**Comment by pimeys** [comment]
Not at all! If you have a working system I want to hear how it works for you with InfluxDB and you can also share some insights what were you able to do with the data.

My system with CGM is all together in about four months from now when my Insurance accepts everything. Until that I need to still build the receiver and I don't have real ways of testing the whole pipeline with actual data.

And it's GPL anyways, so go ahead :)

**Comment by pimeys** [comment]
I finally tried this with real data. I was able to connect my Freestyle Libre with my mobile and xDrip+, got some nice graphs in Grafana. There was a small bug which I fixed in this PR.

Is it possible to send insulins and get the estimations visible in Grafana? Also, maybe a bit off-topic, but it seems every first fetch from NFC seems to give `NFC invalid data - try again` but the second try always gives the measurements. I couldn't see anything in the debug console...

**Comment by pimeys** [comment]
Ah, got it. Just disabled the fast measurement option and now it works. This is so cool :)

**Comment by pimeys** [comment]
Another thing I need to investigate is, that sometimes the client doesn't push the measurements to the server without restarting it first. Today I had it two times, once with no measurements and once with only the latest point but nothing in between. Might be that something goes awry with the connection...

File: app/src/main/java/com/eveningoutpost/dexdrip/InfluxDB/InfluxDBUploader.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+        client = new OkHttpClient.Builder()
+                .connectTimeout(CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                .readTimeout(SOCKET_TIMEOUT, TimeUnit.MILLISECONDS)
+                .writeTimeout(SOCKET_TIMEOUT, TimeUnit.MILLISECONDS)
+                .addNetworkInterceptor(new Interceptor() {
+                    @Override
+                    public Response intercept(Chain chain) throws IOException {
+                        HttpUrl url = HttpUrl.parse(dbUri);
+                        String fullPath = url.encodedPath() + chain.request().url().encodedPath();
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/InfluxDB/InfluxDBUploader.java]: To get this to work locally, I had to filter on a double leading //

`String fullPath = (url.encodedPath() + chain.request().url().encodedPath()).replaceFirst("^//","/");`

**Comment by jamorham** [comment]
I am making some progress with this and have started putting together what I think will be a good version.

The bug in 2.5 version of the java client library makes it a bit messier to commit due to having to bundle the snapshot version which I'd prefer not to do if I can avoid it. Currently it is adding 200KB to the apk size but I think I can improve on that once I'm confident everything is working.

**Comment by pimeys** [comment]
If you're able to write data to InfluxDB and have the latest Grafana communicating with the database, here's the dashboard[0] I've been able to build with the current data. To see all the panels you need Grafana 4.1, InfluxDB 1.x and the Grafana histogram[1] plugin. I have the glucose graph with alerts outside my target levels, a delta graph which is handy for seeing the current trend, a histogram to see how the different values are spread throughout the timespan and the current estimated GHbA1c.

Remember, if you run this on a public server, https+basic auth+fail2ban to block everybody failing to authenticate. I wouldn't trust the security of InfluxDB alone, usually it's not run on a public server.

[0] https://gist.github.com/pimeys/e3a99763aea756f8712abd5dbe19ebdf
[1] https://grafana.net/plugins/mtanda-histogram-panel

**Comment by jamorham** [comment]
Thanks very much for doing this. I have your diabetes panel up. I'm assuming it is mmol only..

 Is there a way we can make it support mg/dl as well? I would also like to send treatment data to Influxdb and would welcome your assistance in some useful ways to display that data. I would also be interested in including the panel in the source tree for people to use.

Using your submitted PR I have produced a version which I think implements the Influxdb support cleanly. The entire cloud uploading system was due for a rework and I've done this at the same time. The commit is here:

https://github.com/NightscoutFoundation/xDrip/commit/5032b2935e57154df282da10a0c2376173ad17fa

Treatments are not yet included but I will get to that soon. Hopefully the new structure using `UploaderQueue` makes sense and avoids painful changes to active android schemas which are unfortunately not downgrade compatible.

**Comment by pimeys** [comment]
You can use mg/dl. The uploader sends both values, just select the value you want in the panel editor or edit the json.

**Comment by jamorham** [comment]
This has now been merged and I put some documentation and credit from this PR in to a simple readme notes documentation file: 

https://github.com/NightscoutFoundation/xDrip/blob/master/Documentation/grafana/Grafana-Notes.md

**Pull request closed without merging**